File: learn/src/secondary/hidden_probe.py
#!/usr/bin/env python3
"""Probe MLP hidden layer dimensionality.

Loads a trained model, runs the scalar→hidden mapping on all atoms,
captures hidden activations, and computes eigenspectrum.  Answers:
how many of the hidden dimensions does the model actually use?

Also reports per-kernel weight statistics: which kernels get weight,
which are suppressed, which flip sign across atoms.

Usage:
    cd learn/src
    python -m secondary hidden_probe --config calibration.toml --run ../runs/azimuthal_scalars_only_50
"""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np
import torch

from .loader import setup_sdk
from mutation_set.config import load_config, Config
from mutation_set.model import ShieldingT2Model
from mutation_set.dataset import CalibrationDataset
from mutation_set.kernels import KernelLayout


def run(cfg: Config, run_dir: str, max_proteins: int = 0):
    run_path = Path(run_dir)
    header = json.loads((run_path / "header.json").read_text())

    n_scalars = header["n_scalar_features"]
    n_kernels = header["n_kernels"]
    kernel_names = header["kernel_names"]

    print(f"Model: {n_kernels} kernels, {n_scalars} scalars, "
          f"hidden={cfg.model.hidden_mix}")

    # Load model
    model = ShieldingT2Model(cfg, n_scalars, n_kernels)
    state = torch.load(run_path / "best_model.pt",
                       map_location="cpu", weights_only=True)
    model.load_state_dict(state)
    model.eval()

    # Load data matching the model's training config.
    # Override kernel layout to match what the model was trained with.
    from dataclasses import replace as dc_replace
    train_calcs = header.get("per_ring_calcs",
                              ["bs", "hm", "chi", "pq", "hm_H", "disp_chi"])
    cfg_matched = dc_replace(cfg, kernels=dc_replace(
        cfg.kernels, per_ring_calculators=train_calcs))
    layout = KernelLayout.from_config(cfg_matched)
    assert layout.n_kernels == n_kernels, \
        f"Layout {layout.n_kernels} != model {n_kernels}"

    from mutation_set.dataset import list_proteins, _build_one
    proteins = list_proteins(cfg.paths.features)
    if max_proteins > 0:
        proteins = proteins[:max_proteins]
    elif header.get("n_proteins", 0) > 0:
        proteins = proteins[:header["n_proteins"]]

    all_s, all_k = [], []
    for pid in proteins:
        result, err = _build_one(cfg.paths.features / pid, cfg_matched, layout)
        if result is not None:
            all_s.append(result["scalars"])
            all_k.append(result["kernels"])

    scalars = torch.tensor(np.vstack(all_s), dtype=torch.float32)
    kernels = torch.tensor(np.vstack(all_k), dtype=torch.float32)
    N = len(scalars)
    print(f"Loaded: {N} atoms")

    # Hook hidden layers
    hidden_outputs = {}

    def make_hook(name):
        def hook(module, inp, out):
            hidden_outputs[name] = out.detach()
        return hook

    # FullyConnectedNet layers: linear, act, linear, act, linear
    # We want the activations after each nonlinearity
    hooks = []
    for i, layer in enumerate(model.mixing.mlp):
        h = layer.register_forward_hook(make_hook(f"layer_{i}"))
        hooks.append(h)

    # Forward pass
    with torch.no_grad():
        relative = model.mixing.mlp(scalars)
        magnitude = kernels.norm(dim=-1)
        gate = magnitude / (magnitude + model.mixing.gate_threshold)
        import math
        weights = (relative * gate / math.sqrt(n_kernels)).numpy()
        relative = relative.numpy()

    for h in hooks:
        h.remove()

    # Analyse each captured layer
    print(f"\n{'=' * 60}")
    print("HIDDEN LAYER DIMENSIONALITY")
    print(f"{'=' * 60}")

    for name in sorted(hidden_outputs.keys()):
        act = hidden_outputs[name].numpy()
        if act.ndim != 2 or act.shape[1] < 2:
            continue
        _analyse(act, f"Layer {name} ({act.shape[1]} units)")

    _analyse(weights, f"Kernel weights, gated ({weights.shape[1]} kernels)")
    _analyse(relative, f"Relative weights, pre-gate ({relative.shape[1]} kernels)")

    # Per-kernel weight stats
    print(f"\n{'=' * 60}")
    print("PER-KERNEL WEIGHT STATISTICS (gated)")
    print(f"{'=' * 60}\n")
    print(f"  {'Kernel':25s}  {'mean|w|':>8s}  {'std|w|':>8s}  "
          f"{'%pos':>6s}  {'%active':>8s}")
    print(f"  {'-'*65}")

    stats = []
    for k in range(n_kernels):
        w = weights[:, k]
        stats.append((
            kernel_names[k],
            float(np.mean(np.abs(w))),
            float(np.std(np.abs(w))),
            float((w > 0).mean()),
            float((np.abs(w) > 0.001).mean()),
        ))

    stats.sort(key=lambda x: x[1], reverse=True)
    for name, ma, sa, fp, fa in stats[:25]:
        print(f"  {name:25s}  {ma:8.4f}  {sa:8.4f}  {fp:6.1%}  {fa:8.1%}")

    print(f"\n  Kernels with mean|w| > 0.01: "
          f"{sum(1 for _, ma, *_ in stats if ma > 0.01)}")
    print(f"  Kernels with mean|w| > 0.001: "
          f"{sum(1 for _, ma, *_ in stats if ma > 0.001)}")
    print(f"  Kernels effectively dead (mean|w| < 0.0001): "
          f"{sum(1 for _, ma, *_ in stats if ma < 0.0001)}")


def _analyse(mat: np.ndarray, label: str):
    """Eigenspectrum of an (N, D) matrix."""
    mat_c = mat - mat.mean(axis=0, keepdims=True)
    cov = mat_c.T @ mat_c / len(mat_c)
    eig = np.linalg.eigvalsh(cov)[::-1]
    eig = np.maximum(eig, 0)
    total = eig.sum()
    if total < 1e-12:
        print(f"\n  {label}: all zeros")
        return
    cumvar = np.cumsum(eig) / total
    dim90 = int(np.searchsorted(cumvar, 0.90)) + 1
    dim95 = int(np.searchsorted(cumvar, 0.95)) + 1
    dim99 = int(np.searchsorted(cumvar, 0.99)) + 1
    active = int((eig > eig[0] * 0.01).sum())

    print(f"\n  {label}:")
    print(f"    Total dims: {mat.shape[1]}")
    print(f"    Active (>1% of max eig): {active}")
    print(f"    90% variance: {dim90} dims")
    print(f"    95% variance: {dim95} dims")
    print(f"    99% variance: {dim99} dims")

File: learn/R/hidden_probe.R
#!/usr/bin/env Rscript
# hidden_probe.R — Probe the MLP hidden layer dimensionality.
#
# Runs the trained MLP's scalar→hidden mapping on all atoms, captures
# the hidden activations, and computes their eigenspectrum.  Answers:
# how many of the 64 hidden dimensions is the model actually using?
#
# Also extracts the final kernel weights per atom and analyses their
# structure: how many distinct weighting patterns exist?
#
# Requires: a trained model checkpoint + the R export data.
# The heavy lifting (forward pass) is done in Python via reticulate.
#
# Usage:
#   cd learn/R
#   Rscript hidden_probe.R ../src/output/secondary/r_export ../runs/azimuthal_scalars_only_50

args <- commandArgs(trailingOnly = TRUE)
source("common.R")

if (length(args) < 2) {
  stop("Usage: Rscript hidden_probe.R <export_dir> <run_dir>")
}
export_dir <- args[1]
run_dir    <- args[2]

library(reticulate)

# ── Run the forward pass in Python to get hidden activations ─────────
cat("Running MLP forward pass via Python...\n")

py_run_string(sprintf("
import sys, json, torch, numpy as np
sys.path.insert(0, '../src')
sys.path.insert(0, '../../python')

from mutation_set.config import load_config
from mutation_set.model import ShieldingT2Model

# Load config and model
cfg = load_config('../src/calibration.toml')
header = json.load(open('%s/header.json'))
n_scalars = header['n_scalars']
n_kernels = header['n_kernels']

model = ShieldingT2Model(cfg, n_scalars, n_kernels)
state = torch.load('%s/best_model.pt', map_location='cpu', weights_only=True)
model.load_state_dict(state)
model.eval()

# Load export data
scalars_np = np.load('%s/scalars.npz')['data']
kernels_np = np.load('%s/kernels_norm.npz')['data']
N = len(scalars_np)

scalars_t = torch.tensor(scalars_np, dtype=torch.float32)
kernels_t = torch.tensor(kernels_np, dtype=torch.float32)

# Hook into the MLP hidden layers
hidden_activations = {}
def hook_fn(name):
    def fn(module, input, output):
        hidden_activations[name] = output.detach().numpy()
    return fn

# The MLP is model.mixing.mlp — FullyConnectedNet with layers
# Register hooks on each linear layer's output
for i, layer in enumerate(model.mixing.mlp):
    if hasattr(layer, 'weight'):
        layer.register_forward_hook(hook_fn(f'layer_{i}'))

# Forward pass
with torch.no_grad():
    # Get the relative weights before gating
    relative = model.mixing.mlp(scalars_t)
    magnitude = kernels_t.norm(dim=-1)
    gate = magnitude / (magnitude + model.mixing.gate_threshold)
    import math
    weights = relative * gate / math.sqrt(n_kernels)

# Save results
hidden1 = hidden_activations.get('layer_1', np.zeros((N, 1)))
hidden2 = hidden_activations.get('layer_3', np.zeros((N, 1)))
kernel_weights = weights.numpy()
relative_weights = relative.numpy()

np.savez('%s/hidden_probe.npz',
         hidden1=hidden1,
         hidden2=hidden2,
         kernel_weights=kernel_weights,
         relative_weights=relative_weights)
print(f'Hidden1: {hidden1.shape}, Hidden2: {hidden2.shape}')
print(f'Kernel weights: {kernel_weights.shape}')
", run_dir, run_dir, export_dir, export_dir, export_dir))

# ── Load results into R ──────────────────────────────────────────────
np <- import("numpy")
probe <- np$load(sprintf("%s/hidden_probe.npz", export_dir), allow_pickle = TRUE)

hidden1  <- probe$f[["hidden1"]]
hidden2  <- probe$f[["hidden2"]]
k_weights <- probe$f[["kernel_weights"]]
rel_weights <- probe$f[["relative_weights"]]

cat(sprintf("Hidden layer 1: %d atoms × %d units\n", nrow(hidden1), ncol(hidden1)))
cat(sprintf("Hidden layer 2: %d atoms × %d units\n", nrow(hidden2), ncol(hidden2)))
cat(sprintf("Kernel weights: %d atoms × %d kernels\n", nrow(k_weights), ncol(k_weights)))

# ── Eigenspectrum of hidden activations ──────────────────────────────

analyse_dim <- function(mat, label) {
  mat_c <- scale(mat, center = TRUE, scale = FALSE)
  cov_mat <- crossprod(mat_c) / nrow(mat_c)
  eig <- eigen(cov_mat, symmetric = TRUE, only.values = TRUE)$values
  eig <- pmax(eig, 0)
  cumvar <- cumsum(eig) / sum(eig)
  dim90 <- which(cumvar >= 0.90)[1]
  dim95 <- which(cumvar >= 0.95)[1]
  dim99 <- which(cumvar >= 0.99)[1]
  active <- sum(eig > max(eig) * 0.01)  # dims > 1% of max eigenvalue
  cat(sprintf("\n%s:\n", label))
  cat(sprintf("  Total dims: %d\n", ncol(mat)))
  cat(sprintf("  Active dims (>1%% of max eig): %d\n", active))
  cat(sprintf("  90%% variance: %d dims\n", dim90))
  cat(sprintf("  95%% variance: %d dims\n", dim95))
  cat(sprintf("  99%% variance: %d dims\n", dim99))

  data.frame(rank = seq_along(eig), eigenvalue = eig, cumvar = cumvar,
             label = label, stringsAsFactors = FALSE)
}

e1 <- analyse_dim(hidden1, "Hidden layer 1")
e2 <- analyse_dim(hidden2, "Hidden layer 2")
ew <- analyse_dim(k_weights, "Kernel weights (gated)")
er <- analyse_dim(rel_weights, "Relative weights (pre-gate)")

all_eigen <- rbind(e1, e2, ew, er)
all_eigen <- all_eigen[all_eigen$rank <= 30, ]

p1 <- ggplot(all_eigen, aes(x = rank, y = cumvar, colour = label)) +
  geom_line(linewidth = 0.8) +
  geom_point(size = 1.5) +
  geom_hline(yintercept = 0.90, linetype = "dashed", colour = "grey50") +
  geom_hline(yintercept = 0.95, linetype = "dotted", colour = "grey50") +
  labs(title = "MLP hidden space dimensionality",
       subtitle = "How many dimensions does the model actually use?",
       x = "Eigenvalue rank", y = "Cumulative variance",
       colour = NULL) +
  coord_cartesian(ylim = c(0, 1))

save_fig(p1, "wb_hidden_probe", width = 10, height = 6)

# ── Per-kernel weight statistics ─────────────────────────────────────
cat("\n\nPer-kernel weight statistics (gated):\n")
cat(sprintf("  %-25s  %8s  %8s  %8s  %8s\n",
            "Kernel", "mean|w|", "std|w|", "frac>0", "frac_active"))

knames <- read.csv(sprintf("%s/kernel_names.csv", export_dir),
                    stringsAsFactors = FALSE)
weight_stats <- list()
for (k in seq_len(ncol(k_weights))) {
  w <- k_weights[, k]
  weight_stats[[k]] <- data.frame(
    kernel = knames$name[k],
    mean_abs = mean(abs(w)),
    std_abs = sd(abs(w)),
    frac_positive = mean(w > 0),
    frac_active = mean(abs(w) > 0.001),
    stringsAsFactors = FALSE
  )
}
ws_df <- do.call(rbind, weight_stats)
ws_df <- ws_df[order(-ws_df$mean_abs), ]
for (i in seq_len(min(20, nrow(ws_df)))) {
  r <- ws_df[i, ]
  cat(sprintf("  %-25s  %8.4f  %8.4f  %8.2f  %8.2f\n",
              r$kernel, r$mean_abs, r$std_abs, r$frac_positive, r$frac_active))
}

cat("\n══════════════════════════════════════════════════════════\n")
cat("  Done. Figure: wb_hidden_probe.pdf\n")
cat("══════════════════════════════════════════════════════════\n")
